internal/jobcontainers/process.go

```diff
@@ -0,0 +1,228 @@
+	<-p.waitBlock
+	return p.waitError
+}
+
+// Start starts the job object process
+func (p *JobProcess) Start() error {
+	return p.cmd.Start()
+}
+
+// This should only be called once.
+func (p *JobProcess) waitBackground(ctx context.Context) {
+	log.G(ctx).WithField("pid", p.Pid()).Debug("waitBackground for JobProcess")
+
+	// Wait for process to get signaled/exit/terminate/.
+	err := p.cmd.Wait()
+
+	// Wait closes the stdio pipes so theres no need to later on.
+	p.stdioLock.Lock()
+	p.stdin = nil
+	p.stdout = nil
+	p.stderr = nil
+	p.stdioLock.Unlock()
+
+	p.closedWaitOnce.Do(func() {
+		p.waitError = err
+		close(p.waitBlock)
+	})
+}
+
+// ExitCode returns the exit code of the process.
+func (p *JobProcess) ExitCode() (int, error) {
+	p.procLock.Lock()
+	defer p.procLock.Unlock()
+
+	if !p.exited() {
+		return -1, errors.New("process has not exited")
+	}
+	return p.cmd.ProcessState.ExitCode(), nil
+}
+
+// Pid returns the processes PID
+func (p *JobProcess) Pid() int {
+	if process := p.cmd.Process; process != nil {
+		return process.Pid
+	}
+	return 0
+}
+
+// Close cleans up any state associated with the process but does not kill it.
+func (p *JobProcess) Close() error {
+	p.stdioLock.Lock()
+	if p.stdin != nil {
+		p.stdin.Close()
+		p.stdin = nil
+	}
+	if p.stdout != nil {
+		p.stdout.Close()
+		p.stdout = nil
+	}
+	if p.stderr != nil {
+		p.stderr.Close()
+		p.stderr = nil
+	}
+	p.stdioLock.Unlock()
+
+	p.closedWaitOnce.Do(func() {
+		p.waitError = hcs.ErrAlreadyClosed
+		close(p.waitBlock)
+	})
+	return nil
+}
+
+// Kill kills the running process. Go calls TerminateProcess under the hood.
+func (p *JobProcess) Kill(ctx context.Context) (bool, error) {
+	log.G(ctx).WithField("pid", p.Pid()).Debug("killing job process")
+
+	p.procLock.Lock()
+	defer p.procLock.Unlock()
+
+	if p.exited() {
+		return false, errors.New("kill not sent. process already exited")
+	}
+
+	if p.cmd.Process != nil {
+		if err := p.cmd.Process.Kill(); err != nil {
+			return false, err
+		}
+	}
+	return true, nil
+}
+
+func (p *JobProcess) exited() bool {
+	if p.cmd.ProcessState == nil {
+		return false
+	}
+	return p.cmd.ProcessState.Exited()
+}
+
+// signalProcess sends the specified signal to a process. We can't use GenerateConsoleCtrlEvent since
+// that only supports CTRL_C_EVENT and CTRL_BREAK_EVENT. Instead to handle an arbitrary signal,
+// we open a CtrlRoutine thread inside the target process and give it the specified signal to handle.
+//
+// Note: This is a hack which is not officially supported.
+func signalProcess(pid uint32, signal int) error {
+	hProc, err := windows.OpenProcess(winapi.PROCESS_ALL_ACCESS, true, pid)
+	if err != nil {
+		return errors.Wrap(err, "failed to open process")
+	}
+	defer windows.Close(hProc)
+
+	k32, err := windows.LoadLibrary("kernel32.dll")
+	if err != nil {
+		return errors.Wrap(err, "failed to load kernel32 library")
+	}
+	defer windows.Close(k32)
+
+	proc, err := windows.GetProcAddress(k32, "CtrlRoutine")
```

Comment:
This will give us the proc address in the shim process. Do we know that we can rely on the address being the same in the target process? This module won't be ASLR'd?

Reply:
We should have a comment explaining why this is safe.

Reply:
I did a little reading and it sounds like ASLR will randomize locations each boot, but all loads of the same library in the same boot session will have the same address. So I think we're okay, but should document this.

Reply:
Yep you're spot on. I was curious to hear your thoughts on this, didn't know how much we wanted to document this given it's hacky nature.

Reply:
We should at least make it clear to a reader that it will actually work. Otherwise it's really easy to read this in a few years when hunting some obscure bug and think this could be the cause.

Reply:
Good point. Let me know if what I added looks sufficient